app/assets/javascripts/govuk_publishing_components/components/search-autocomplete.js

```diff
@@ -0,0 +1,318 @@
+/* global fetch */
+
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {}
+window.GOVUK.Modules.GovukAutocomplete = window.GOVUKFrontend.Autocomplete;
+
+(function (Modules) {
+  'use strict'
+
+  class GemSearchAutocomplete {
+    constructor ($module) {
+      this.$module = $module
+
+      this.numberSuggestions = $module.getAttribute('data-display-number-suggestions') || 5
+      this.throttleDelayTime = $module.getAttribute('data-throttle-delay-time') || 0
+      this.submitOnSelect = $module.getAttribute('data-submit-on-select') || false
+      this.baseClass = $module.getAttribute('data-base-class') || 'gem-c-search-autocomplete'
+      this.source = $module.getAttribute('data-source')
+    }
+
+    init () {
+      this.onUpdate = this.handleUpdate
+      this.onSubmit = () => { }
+
+      this.expanded = false
+      this.loading = false
+
+      this.$module.insertAdjacentHTML('beforeend', '<ul class="gem-c-search-autocomplete__result-list govuk-body" role="listbox"></ul><div aria-atomic="true" aria-live="polite" role="status" class="govuk-visually-hidden">No results.</div>')
+
+      this.$resultList = this.$module.querySelector('ul')
+      this.$resultList.setAttribute('role', 'listbox')
+      this.$resultList.id = `${this.baseClass}-result-list-${this.$module.getAttribute('data-id-postfix')}`
+
+      this.$resultList.addEventListener('mousedown', this.handleResultMouseDown)
+      this.$resultList.addEventListener('click', this.handleResultClick.bind(this))
+
+      this.$input = this.$module.querySelector('input')
+      this.$input.setAttribute('role', 'combobox')
+      this.$input.setAttribute('autocomplete', 'off')
+      this.$input.setAttribute('aria-autocomplete', 'list')
+      this.$input.setAttribute('aria-haspopup', 'listbox')
+      this.$input.setAttribute('aria-expanded', 'false')
+      this.$input.setAttribute('aria-owns', this.$resultList.id)
+      this.$input.setAttribute('autocapitalize', 'off')
+      this.$input.setAttribute('spellcheck', 'false')
+
+      this.$input.addEventListener('input', this.handleInput.bind(this))
+      this.$input.addEventListener('keydown', this.handleKeyDown.bind(this))
+      this.$input.addEventListener('keyup', this.handleKeyUp.bind(this))
+      this.$input.addEventListener('focus', this.handleFocus.bind(this))
+      this.$input.addEventListener('blur', this.handleBlur.bind(this))
+
+      document.body.addEventListener('click', this.handleDocumentClick.bind(this))
+
+      this.createAssistiveHint()
+
+      this.$liveRegion = this.$module.querySelector('[role="status"]')
+
+      this.value = ''
+      this.searchCounter = 0
+      this.results = []
+      this.selectedIndex = -1
+      this.selectedResult = null
+      this.throttled = false
+
+      this.normaliseSource()
+    }
+
+    normaliseSource () {
+      if (Array.isArray(this.source.split(',').map(item => item.trim()))) {
```

Comment:
resolving this as removed rendundant map and other point is covered with next PR comment

Reply:
The exception handling you've replaced it with seems undesirable as we're using exceptions for control flow - I'd image you'd have things clearer if you had a sourcesUrl option. You probably want something that flags the JSON ones as dummy demo options since they can't actually be used for the component in a live setting.